axelor-production/src/main/java/com/axelor/apps/production/service/SolDetailsBomUpdateServiceImpl.java

```diff
@@ -0,0 +1,148 @@
+package com.axelor.apps.production.service;
+
+import com.axelor.apps.base.AxelorException;
+import com.axelor.apps.base.db.repo.ProductRepository;
+import com.axelor.apps.production.db.BillOfMaterialLine;
+import com.axelor.apps.production.db.SaleOrderLineDetails;
+import com.axelor.apps.production.db.repo.BillOfMaterialLineRepository;
+import com.axelor.apps.production.db.repo.BillOfMaterialRepository;
+import com.axelor.apps.production.db.repo.SaleOrderLineDetailsRepository;
+import com.axelor.apps.sale.db.SaleOrderLine;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.commons.collections.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SolDetailsBomUpdateServiceImpl implements SolDetailsBomUpdateService {
+
+  protected final BillOfMaterialRepository billOfMaterialRepository;
+  protected final BillOfMaterialLineRepository billOfMaterialLineRepository;
+  protected final BomLineCreationService bomLineCreationService;
+  private static final Logger logger =
+      LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  @Inject
+  public SolDetailsBomUpdateServiceImpl(
+      BillOfMaterialRepository billOfMaterialRepository,
+      BillOfMaterialLineRepository billOfMaterialLineRepository,
+      BomLineCreationService bomLineCreationService) {
+    this.billOfMaterialRepository = billOfMaterialRepository;
+    this.billOfMaterialLineRepository = billOfMaterialLineRepository;
+    this.bomLineCreationService = bomLineCreationService;
+  }
+
+  @Override
+  @Transactional(rollbackOn = Exception.class)
+  public void updateSolDetailslWithBillOfMaterial(SaleOrderLine saleOrderLine)
```

Comment:
please split this method, they are too many nested for and if

also this method is very similar to `updateSolWithBillOfMaterial(SaleOrderLine saleOrderLine)`, please check if we can factorize some logic